File: design_patterns/C++/Structural/adapter.cxx
#include <iostream>

class Lion {
  public:
    virtual void roar() = 0;
};

class AfricanLion: public Lion {
  public:
    virtual void roar() = 0;
};

class AsianLion: public Lion {
  public:
    void roar() {}
};

class Hunter {
  public:
    void hunt(Lion* lion) {}
};

class WildDog {
  public:
    void bark() {}
};

class WildDogAdapter: public Lion {
  protected:
    WildDog* dog = nullptr;

  public:
    WildDogAdapter(WildDog* dog) {
        this->dog = dog;
    }

    void roar() {
        this->dog->bark();
    }
};

int main() {
    auto wildDog = new WildDog();
    auto wildDogAdapter = new WildDogAdapter(wildDog);
    auto hunter = new Hunter();
    hunter->hunt(wildDogAdapter);
    return 0;
}

File: design_patterns/Python/Structural/adapter.py
class Lion:

    def roar(self):
        raise NotImplementedError


class AfricanLion(Lion):

    def roar():
        pass


class AsianLion(Lion):

    def roar():
        pass


class Hunter:

    def hunt(self, lion):
        pass


class WildDog:

    def bark(self):
        pass


class WildDogAdapter(Lion):

    def __init__(self, dog):
        self.dog = dog

    def roar(self):
        self.dog.bark()


if __name__ == '__main__':

    wild_dog = WildDog()
    wild_dog_adapter = WildDogAdapter(wild_dog)
    hunter = Hunter()
    hunter.hunt(wild_dog_adapter)
